docs/virtual-package.md

```diff
@@ -0,0 +1,9 @@
+# Composer Virtual Packages
+
+Virtual packages are a way to specify the dependency on an implementation of an interface-only repository without forcing a specific implementation. For Httplug, the virtual package is called `php-http/client-implementation`. 
+
+There is no project registered with that name. However, all client implementations including client adapters for Httplug use the `provide` section to tell composer that they do provide the client-implementation.
+
+When *writing a reusable library*, the `require` section should only mention `php-http/client-implementation`, while the `require-dev` section needs to specify a concrete implementation in most cases, in order for the package to be installable on its own for development and testing.
+
+When *using a reusable library* in an application, the `require` section needs to include a concrete implementation in order for the application to install. Missing to do that will lead to composer reporting a missing package.
```

Comment:
Okay, if something, then a gif animation, because it looks good :stuck_out_tongue_winking_eye: 

For copy pasters we can have the textual error as well.